File: engine/src/market/Updates.java
package market;

import java.util.ArrayList;
import java.util.List;

public class Updates {
    private List<FeedbackUpdate> feedbacks;
    private List<OrderUpdate> orders;
    private List<StoreUpdate> stores;

    Updates(){
        reset();
    }

    public void reset(){
        this.feedbacks = new ArrayList<>();
        this.orders = new ArrayList<>();
        this.stores = new ArrayList<>();
    }

    public void addFeedbackUpdate(FeedbackUpdate feedbackUpdate){
        feedbacks.add(feedbackUpdate);
    }

    public void addOrderUpdate(OrderUpdate orderUpdate){
        orders.add(orderUpdate);
    }

    public void addStoreUpdate(StoreUpdate storeUpdate){
        stores.add(storeUpdate);
    }

    public int getFeedbacksVersion() {
        return feedbacks.size();
    }

    public int getOrderVersion() {
        return orders.size();
    }

    public int getNewStoreVersion() {
        return stores.size();
    }

    public List<FeedbackUpdate> getFeedbackEntries(int fromIndex){
        if (fromIndex < 0 || fromIndex > feedbacks.size()) {
            fromIndex = 0;
        }
        return feedbacks.subList(fromIndex, feedbacks.size());
    }
    public List<OrderUpdate> getOrdersEntries(int fromIndex){
        if (fromIndex < 0 || fromIndex > orders.size()) {
            fromIndex = 0;
        }
        return orders.subList(fromIndex, orders.size());
    }

    public List<StoreUpdate> getNewStoreEntries(int fromIndex){
        if (fromIndex < 0 || fromIndex > stores.size()) {
            fromIndex = 0;
        }
        return stores.subList(fromIndex, stores.size());
    }

    public static class FeedbackUpdate{
        String customerName;
        String zone;
        int rating;
        String comment;
        String date;

        public FeedbackUpdate(String customerName, String zone, int rating, String comment, String date) {
            this.customerName = customerName;
            this.zone = zone;
            this.rating = rating;
            this.comment = comment;
            this.date = date;
        }
    }

    public static class OrderUpdate{
        int orderSerialNumber;
        int subOrderSerialNumber;
        String customerName;
        int numberOfProductsTypes;
        double productsPrice;
        double deliveryPrice;

        public OrderUpdate(int orderSerialNumber, int subOrderSerialNumber, String customerName, int numberOfProductsTypes, double productsPrice, double deliveryPrice) {
            this.orderSerialNumber = orderSerialNumber;
            this.subOrderSerialNumber = subOrderSerialNumber;
            this.customerName = customerName;
            this.numberOfProductsTypes = numberOfProductsTypes;
            this.productsPrice = productsPrice;
            this.deliveryPrice = deliveryPrice;
        }
    }

    public static class StoreUpdate{
        String ownerName;
        String storeName;
        Coordinates storeLocation;
        String zone;
        int storeNumberOfProducts;
        int zoneNumberOfProducts;

        public StoreUpdate(String ownerName, String storeName, String zone, Coordinates storeLocation, int storeNumberOfProducts, int zoneNumberOfProducts) {
            this.ownerName = ownerName;
            this.storeName = storeName;
            this.storeLocation = storeLocation;
            this.zone = zone;
            this.storeNumberOfProducts = storeNumberOfProducts;
            this.zoneNumberOfProducts = zoneNumberOfProducts;
        }
    }
}

File: MarketWebApp/web/resource/updates.js
function addFeedbackUpdate(index, feedback) {
    var currentText = $('.update-text').text();
    var feedbackDetails = "NEW FEEDBACK !!! customer : " + feedback.customer + ",  rating : " + feedback.rating + "/5" +
        ",  comment : " + feedback.comment + "        ";
    $('.update-text').text(currentText + feedbackDetails);
}

function addStoreUpdate(index, store) {
    var currentText = $('.update-text').text();
    var storeDetails = "NEW STORE !!! store name : " + store.storeName + ",  owner : " + store.ownerName + ",  zone : "
        + store.zone + ",  products : " + store.storeNumberOfProducts + "/" + store.zoneNumberOfProducts + "        ";
    $('.update-text').text(currentText + storeDetails);
}

function addOrderUpdate(index, order) {
    var currentText = $('.update-text').text();
    var orderDetails = "NEW ORDER !!! Order no' : " + order.orderSerialNumber + "." + order.subOrderSerialNumber +
        + ",  Customer : " + order.customerName + ",  Products : " + order.numberOfProductsTypes + ",  Products price : " + order.productsPrice
        + ",  Delivery price :" + order.deliveryPrice;
    $('.update-text').text(currentText + orderDetails);
}

function printAndGetUpdates(updatesAndVersions) {
    if(updatesAndVersions.feedbacks.length == 0 && updatesAndVersions.stores.length == 0 && updatesAndVersions.orders.length == 0){
        return;
    } else {
        setAllData();
        $.each(updatesAndVersions.feedbacks || [], addFeedbackUpdate);
        $.each(updatesAndVersions.stores || [], addStoreUpdate);
        $.each(updatesAndVersions.orders || [], addOrderUpdate);
    }
}

function getUpdates() {
    $.ajax({
        method: 'GET',
        url: 'update',
        dataType: 'JSON',
        data: {
            updateOrderVersion: updateOrderVersion,
            updateFeedbackVersion : updateFeedbackVersion,
            updateNewStoreVersion: updateNewStoreVersion
        },
        success: function (updatesAndVersions){
            updateOrderVersion = updatesAndVersions.orderVersion;
            updateFeedbackVersion = updatesAndVersions.feedbackVersion;
            updateNewStoreVersion = updatesAndVersions.newStoreVersion;
            printAndGetUpdates(updatesAndVersions);
        },
        error: function (){
            console.log("Failed to get updates");
        }
    });
}